### PR TITLE
rules_rust: enable pipelined compilation.

### PR DESCRIPTION
**Pull request by gigaroby**
Pipelined compilation allows better parallelism during builds as it
allows libraries to generate lightweight metadata files to unlock other
depencies. These metadata files (.rmeta) can only be used to unlock
library -> library dependencies and do not affect builds in any other
way. This is currently the default in cargo:
https://internals.rust-lang.org/t/evaluating-pipelined-rustc-compilation/10199

### CODE REVIEW

**Comment by dtolnay** [comment]
Mentioning @jsgf since you wrote tips on how to implement this in https://github.com/bazelbuild/rules_rust/issues/228#issuecomment-1100788353 based on Buck v2.

File: rust/private/providers.bzl [hunk 1]
```diff
@@ -49,6 +50,7 @@ DepInfo = provider(
         "transitive_build_infos": "depset[BuildInfo]",
         "transitive_crate_outputs": "depset[File]: All transitive crate outputs.",
         "transitive_crates": "depset[CrateInfo]",
+        "transitive_metadata_outputs": "depset[File]: All transitive crate metadata outputs.",
```
Comment (scentini) [inline rust/private/providers.bzl:53]: I believe this description is not correct anymore, as we're now adding also `.rlib`s from the dependencies that don't have metadata enabled. Maybe something along the lines of `Transitive outputs for pipelined compilation` fits better?

File: rust/private/rustc.bzl [hunk 2]
```diff
@@ -1304,17 +1448,34 @@ def _get_dir_names(files):
...
     """Adds link flags to an Args object reference
 
     Args:
         args (Args): An arguments object reference
         dep_info (DepInfo): The current target's dependency info
         force_all_deps_direct (bool, optional): Whether to pass the transitive rlibs with --extern
             to the commandline as opposed to -L.
+        link_to_metadata:
     """
+    # print(" link_to_metadata ", link_to_metadata)
```
Comment (scentini) [inline rust/private/rustc.bzl:1461]: Leftover from debugging?

File: rust/private/rustc.bzl [hunk 3]
```diff
@@ -1304,17 +1448,34 @@ def _get_dir_names(files):
...
-def add_crate_link_flags(args, dep_info, force_all_deps_direct = False):
+def add_crate_link_flags(args, dep_info, force_all_deps_direct = False, link_to_metadata = False):
     """Adds link flags to an Args object reference
 
     Args:
         args (Args): An arguments object reference
         dep_info (DepInfo): The current target's dependency info
         force_all_deps_direct (bool, optional): Whether to pass the transitive rlibs with --extern
             to the commandline as opposed to -L.
+        link_to_metadata:
```
Comment (scentini) [inline rust/private/rustc.bzl:1459]: Missing documentation.

File: rust/private/rustc.bzl [hunk 4]
```diff
@@ -1304,17 +1448,34 @@ def _get_dir_names(files):
...
     """
+    # print(" link_to_metadata ", link_to_metadata)
 
-    if force_all_deps_direct:
+    if not link_to_metadata:
```
Comment (scentini) [inline rust/private/rustc.bzl:1463]: This section was a big hard to follow for me. Wdyt about sth like:

```
direct_crates = depset(
    transitive = [dep_info.direct_crates, dep_info.transitive_crates]
  ) if force_all_deps_direct else dep_info.direct_crates

crate_to_link_flags = _crate_to_link_flag_metadata if link_to_metadata else _crate_to_link_flag

args.add_all(direct_crates, uniquify=True, map_each = crate_to_link_flags)

```

File: rust/private/rustc.bzl [hunk 5]
```diff
@@ -1304,17 +1448,34 @@ def _get_dir_names(files):
         dirs[f.dirname] = None
     return dirs.keys()
 
-def add_crate_link_flags(args, dep_info, force_all_deps_direct = False):
+def add_crate_link_flags(args, dep_info, force_all_deps_direct = False, link_to_metadata = False):
```
Comment (scentini) [inline rust/private/rustc.bzl:1451]: I believe this function is not named properly, we add these flags when compiling `rlib`s too, not just for linking actions. While renaming the function itself does not need to be in scope for this PR, could we remove the `link` word from the new parameter? `is_pipelined_compilation`, `use_metadata` come to mind.

File: rust/private/rustdoc.bzl [hunk 6]
```diff
@@ -37,6 +37,7 @@ def _strip_crate_info_output(crate_info):
         aliases = crate_info.aliases,
         # This crate info should have no output
         output = None,
+        metadata = crate_info.metadata,
```
Comment (scentini) [inline rust/private/rustdoc.bzl:40]: Given the comment 2 lines above, probably this crate info should also not have a metadata output?

File: util/process_wrapper/main.rs [hunk 7]
```diff
@@ -40,6 +40,7 @@ fn status_code(status: ExitStatus, was_killed: bool) -> i32 {
 fn status_code(status: ExitStatus, was_killed: bool) -> i32 {
     // On unix, if code is None it means that the process was killed by a signal.
     // https://doc.rust-lang.org/std/process/struct.ExitStatus.html#method.success
+    // eprintln!("{:?} {}", status, was_killed);
```
Comment (scentini) [inline util/process_wrapper/main.rs:43]: Leftover from debugging?

File: test/unit/force_all_deps_direct/generator.bzl [hunk 8]
```diff
@@ -39,6 +42,12 @@ EOF
         lib_hash = output_hash,
         extension = ".rlib",
     )
+    rust_metadata_name = "{prefix}{name}-{lib_hash}{extension}".format(
```
Comment (scentini) [inline test/unit/force_all_deps_direct/generator.bzl:45]: This rule has a different purpose than testing pipelining, could we revert the changes here and clone the file under test/unit/pipelined_compilation and make the changes there?

Addidionally, could we have a rule like this one, that doesn't support pipelined compilation, and check that a 
rust_library_a -> custom_rule_without_compilation -> rust_library_b works?
We could have two different tests:
a) with enabled pipelining, the `custom_rule_without_compilation` action should still compile with the rust_library_b`.rmeta` input, even though `custom_rule_without_compilation` will not produce `.rmeta` output itself.
b) with enabled pipelining, the `rust_library_a` target will still compile successfully, by using the `custom_rule_without_compilation` `.rlib` output.

File: test/unit/pipelined_compilation/pipelined_compilation_test.bzl [hunk 9]
```diff
@@ -0,0 +1,125 @@
+"""Unittests for rust rules."""
```
Comment (scentini) [inline test/unit/pipelined_compilation/pipelined_compilation_test.bzl:1]: Thanks for the extensive coverage!
One more test I would like to see is that in the case of 
`rust_library_a` -> `proc_macro` -> `rust_library_b`, when compiling `rust_library_a` object or rmeta file, we do not pass the `rust_library_b` `.rmeta` or `.rlib` as an input (we shouldn't propagate inputs from `proc_macro` depenendencies).
I think https://github.com/bazelbuild/rules_rust/blob/main/test/unit/proc_macro/leaks_deps/proc_macro_does_not_leak_deps.bzl will be a fitting place for such a test, as it tests the same for `.rlib` and native dependencies.

File: rust/private/rustdoc.bzl (hunk 6 above)
Comment (gigaroby) [reply rust/private/rustdoc.bzl:40]: Done.

File: rust/private/rustc.bzl [hunk 10]
```diff
@@ -222,19 +226,35 @@ def collect_deps(
...
+            )
+
+            # If this dependency produces metadata, add it to the metadata outputs.
+            # If it doesn't (for example a custom library that exports crate_info),
+            # we depend on crate_info.output.
+            depend_on = crate_info.metadata
+            if not crate_info.metadata:
+                depend_on = crate_info.output
+
+            transitive_metadata_outputs.append(
```
[the remark this answers is not in the document]
Comment (gigaroby) [reply rust/private/rustc.bzl:240]: Done.

File: rust/private/rustc.bzl (hunk 2 above)
Comment (gigaroby) [reply rust/private/rustc.bzl:1461]: Removed.

File: rust/private/rustc.bzl (hunk 5 above)
Comment (gigaroby) [reply rust/private/rustc.bzl:1451]: Done.

File: rust/private/rustc.bzl (hunk 3 above)
Comment (gigaroby) [reply rust/private/rustc.bzl:1459]: Done.

File: test/unit/pipelined_compilation/pipelined_compilation_test.bzl (hunk 9 above)
Comment (gigaroby) [reply test/unit/pipelined_compilation/pipelined_compilation_test.bzl:1]: Done.

File: util/process_wrapper/main.rs (hunk 7 above)
Comment (gigaroby) [reply util/process_wrapper/main.rs:43]: Done.

File: test/unit/force_all_deps_direct/generator.bzl (hunk 8 above)
Comment (gigaroby) [reply test/unit/force_all_deps_direct/generator.bzl]: Done.

File: test/unit/pipelined_compilation/pipelined_compilation_test.bzl [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+    else:
+        asserts.true(env, seen_wrapper_rlib, "expected dependency on rlib for 'wrapper' but not found")
+
+    asserts.true(env, seen_to_wrap, "expected dependency on metadata for 'to_wrap' but not found")
+
+    return analysistest.end(env)
+
+custom_rule_test = analysistest.make(_custom_rule_test_impl, attrs = {"generate_metadata": attr.bool()}, config_settings = ENABLE_PIPELINING)
+
+def _custom_rule_test(generate_metadata, prefix):
```
Comment (scentini) [inline test/unit/pipelined_compilation/pipelined_compilation_test.bzl:150]: Nit: Looks like our `prefix` is actually a suffix :)

File: test/unit/pipelined_compilation/pipelined_compilation_test.bzl [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+        if "libto_wrap" in mi.path:
+            seen_to_wrap = True
+
+    seen_wrapper_rlib = True
+    for ri in rlib_inputs:
+        if "libwrapper" in ri.path:
+            seen_wrapper_rlib = True
+
+    if ctx.attr.generate_metadata:
+        asserts.true(env, seen_wrapper_metadata, "expected dependency on metadata for 'wrapper' but not found")
```
Comment (scentini) [inline test/unit/pipelined_compilation/pipelined_compilation_test.bzl:140]: Could we also assert the reverse - that we have not seen the `rlib` in the first case and that we have not seen the `rmeta` in the second case?

File: test/unit/pipelined_compilation/pipelined_compilation_test.bzl [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+    else:
+        asserts.true(env, seen_wrapper_rlib, "expected dependency on rlib for 'wrapper' but not found")
+
+    asserts.true(env, seen_to_wrap, "expected dependency on metadata for 'to_wrap' but not found")
+
+    return analysistest.end(env)
+
+custom_rule_test = analysistest.make(_custom_rule_test_impl, attrs = {"generate_metadata": attr.bool()}, config_settings = ENABLE_PIPELINING)
```
Comment (scentini) [inline test/unit/pipelined_compilation/pipelined_compilation_test.bzl:148]: This test tests that in `rust_library -> custom_rule -> rust_library` with enabled pipelining, when we build the first `rust_library` we use the `rmeta` of the last library and, when available, the `.rmeta` of the `custom_rule`.

I believe that when we build the `custom_rule` with `ctx.attr.generate_metadata=False` currently we end up not using the `.rmeta` of the last `rust_library` (see my comment on the `_depend_on_metadata()` function. can we rename this test to `rmeta_is_propagated_through_custom_rule_test`, and then have a second one `rmeta_is_used_when_building_custom_rule_test` that will have the `"wrapper" + prefix` target under test and will check that regardless of the `ctx.attr.generate_metadata`, we use the available metadata from the `rust_library` dependency?

File: rust/private/rustc.bzl [hunk 14]
```diff
@@ -505,6 +522,30 @@ def _disambiguate_libs(actions, toolchain, crate_info, dep_info, use_pic):
...
+        force_depend_on_objects (bool): if set we will not depend on metadata.
+
+    Returns:
+        Whether we can depend on metadata for this crate.
+    """
+    if force_depend_on_objects:
+        return False
+
+    is_lib = crate_info.type in ("rlib", "lib")
+    build_metadata = crate_info.metadata != None
```
Comment (scentini) [inline rust/private/rustc.bzl:546]: I think we should remove this line. Even if the current crate doesn't produce .rmeta, we can still depend on the metadata from its dependencies when building its `.rlib`

File: test/unit/pipelined_compilation/pipelined_compilation_test.bzl (hunk 11 above)
Comment (gigaroby) [reply test/unit/pipelined_compilation/pipelined_compilation_test.bzl:150]: Done.

**Comment by purkhusid** [comment]
@gigaroby I'm curious if you have already used this on a decently sized project? If so, did you see any noticeable difference?

**Comment by durin42** [comment]
> @gigaroby I'm curious if you have already used this on a decently sized project? If so, did you see any noticeable difference?

Building the Rust toolchain using bazel this seems to reduce the wait from 2h to 40 minutes. It's an enormous win!

**Comment by purkhusid** [comment]
Nice! Looking forward to test this on our repo.

**Comment by UebelAndre** [comment]
> > @gigaroby I'm curious if you have already used this on a decently sized project? If so, did you see any noticeable difference?
> 
> Building the Rust toolchain using bazel this seems to reduce the wait from 2h to 40 minutes. It's an enormous win!

What are you building? A a full rust toolchain? So rustc and things? Or do I misunderstand that.

**Comment by durin42** [comment]
> > > @gigaroby I'm curious if you have already used this on a decently sized project? If so, did you see any noticeable difference?
> > 
> > 
> > Building the Rust toolchain using bazel this seems to reduce the wait from 2h to 40 minutes. It's an enormous win!
> 
> What are you building? A a full rust toolchain? So rustc and things? Or do I misunderstand that.

Yes, a full Rust toolchain including LLVM, bootstrapped from an existing beta rustc and clang.

**Review by scentini — APPROVED** [review]

**Comment by purkhusid** [comment]
Nice! @scentini is it possible to cut a release with this feature soon?

**Comment by jsgf** [comment]
@durin42 Presumably it had no effect on the llvm bits of the build?

**Comment by gigaroby** [comment]
@jsgf not unless those parts of LLVM are in rust themselves. This only unlocks parallelism for rlib -> rlib arches in the build graph.
